pkg/kubectl/cmd/create_deployment.go

```diff
@@ -76,25 +135,21 @@ func CreateDeployment(f cmdutil.Factory, cmdOut, cmdErr io.Writer, cmd *cobra.Co
 	if err != nil {
 		return fmt.Errorf("failed to discover supported resources: %v", err)
 	}
+
 	generatorName := cmdutil.GetFlagString(cmd, "generator")
-	// fallback to the old generator if server does not support apps/v1beta1 deployments
-	if generatorName == cmdutil.DeploymentBasicAppsV1Beta1GeneratorName &&
-		!contains(resourcesList, appsv1beta1.SchemeGroupVersion.WithResource("deployments")) {
-		fmt.Fprintf(cmdErr, "WARNING: New deployments generator specified (%s), but apps/v1beta1.Deployments are not available, falling back to the old one (%s).\n",
-			cmdutil.DeploymentBasicAppsV1Beta1GeneratorName, cmdutil.DeploymentBasicV1Beta1GeneratorName)
-		generatorName = cmdutil.DeploymentBasicV1Beta1GeneratorName
-	}
-	var generator kubectl.StructuredGenerator
-	switch generatorName {
-	case cmdutil.DeploymentBasicAppsV1Beta1GeneratorName:
-		generator = &kubectl.DeploymentBasicAppsGeneratorV1{Name: name, Images: cmdutil.GetFlagStringSlice(cmd, "image")}
-	case cmdutil.DeploymentBasicV1Beta1GeneratorName:
-		generator = &kubectl.DeploymentBasicGeneratorV1{Name: name, Images: cmdutil.GetFlagStringSlice(cmd, "image")}
-	default:
-		return cmdutil.UsageError(cmd, fmt.Sprintf("Generator: %s not supported.", generatorName))
+
+	// It is possible we have to modify the user-provided generator name if
+	// the server does not have support for the requested generator.
+	generatorName = fallbackGeneratorNameIfNecessary(generatorName, resourcesList, cmdErr)
+
+	imageNames := cmdutil.GetFlagStringSlice(cmd, "image")
+	generator, ok := generatorFromName(generatorName, imageNames, deploymentName)
```

Comment:
> also if there's only one invoke here, I don't think it's needed to abstract those logic for function generatorFromName here.

Imagine you were writing unit tests for the original function (on master). In order to test that the correct generator is chosen from a given name, you have to make a series of test cases that stub or mock these things:
 * `cmdutil.Factory`
 * `cobra.Command`
 * two `io.Writer`s *and*
 * the final process of `RunCreateSubcommand`, which involves
   * a fake REST client
   * a fake `APIFactory`
   * a `restclient.Config{}`

That's too much mocking and stubbing. It's painful. In fact, it's so painful that [nobody has written test code for it](https://github.com/kubernetes/kubernetes/blob/584b1139592787638eaba23670354287fe21fe31/pkg/kubectl/cmd/create_deployment_test.go).

Imagine you were writing unit tests for this function on my refactor branch. You can easily verify that the correct generator is chosen from a given name. Just make up three variables (primitives, no less), pass them to `generatorFromName`, and see if the generator is what you expect. There are no side effects or HTTP calls. There's no mocking. Your test becomes a simple correctness check that will never fail unless `generatorFromName` is doing the wrong thing.

Let me know if there's something I've missed. Thanks!